dspace-api/src/main/java/org/dspace/discovery/DiscoverResultIterator.java

```diff
@@ -0,0 +1,144 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.discovery;
+
+import java.io.Serializable;
+import java.sql.SQLException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import org.dspace.core.Context;
+import org.dspace.core.ReloadableEntity;
+
+/**
+ * Implementation of {@link Iterator} to iterate over the discover search result.
+ *
+ * @param <T>  the type of the indexed object
+ * @param <PK> the type of the id of the indexed object
+ * @author Luca Giamminonni (luca.giamminonni at 4science.it)
+ * @author Luca Giamminonni (luca.giamminonni at 4science.it)
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class DiscoverResultIterator<T extends ReloadableEntity, PK extends Serializable> implements Iterator<T> {
```

Comment:
It may be nice to add some basic tests (either Unit or Integration tests) for both `DiscoverResultIterator` and `DiscoverResultItemIterator`, as I'm seeing a **lot** of warnings from our Codecov scanner in these classes saying this code is untested.

I like the design of each of these iterators, but adding automated tests would help ensure they are behaving properly.